specs/agents/tracing-spans-compress.md

```diff
@@ -0,0 +1,211 @@
+This strategy detects consecutive spans that hold the same information (except for the duration)
+and creates a single [composite span](#composite-span).
+
+```
+[                              ]
+GET /users
+ [] [] [] [] [] [] [] [] [] []
+ 10x SELECT FROM users
+```
+
+Two spans are considered to be an exact match if they are of the [same kind](#consecutive-same-kind-compression-strategy) and if their span names are equal:
+- `type`
+- `subtype`
+- `destination.service.resource`
+- `name`
+
+### Consecutive-Same-Kind compression strategy
+
+Another pattern that often occurs is a high amount of alternating queries to the same backend.
+Especially if the individual spans are quite fast, recording every single query is likely to not be worth the overhead.
+
+```
+[                              ]
+GET /users
+ [] [] [] [] [] [] [] [] [] []
+ 10x Calls to mysql
+```
+
+Two spans are considered to be of the same type if the following properties are equal:
+- `type`
+- `subtype`
+- `destination.service.resource`
+
+```java
+boolean isSameKind(Span other) {
+    return type == other.type
+        && subtype == other.subtype
+        && destination.service.resource == other.destination.service.resource
+}
+```
+
+When applying this compression strategy, the `span.name` is set to `Calls to $span.destination.service.resource`.
+The rest of the context, such as the `db.statement` will be determined by the first compressed span, which is turned into a composite span.
+
+
+#### Configuration option `same_kind_compression_max_duration`
+
+Consecutive spans to the same destination that are under this threshold will be compressed into a single composite span.
+This option does not apply to [composite spans](#composite-span).
+This reduces the collection, processing, and storage overhead, and removes clutter from the UI.
+The tradeoff is that the statement information will not be collected. 
+
+|                |          |
+|----------------|----------|
+| Type           | `duration`|
+| Default        | `5ms`    |
+| Dynamic        | `true`   |
+
+### Composite span
+
+Compressed spans don't have a physical span document.
+Instead, multiple compressed spans are represented by a composite span.
+
+#### Data model
+
+The `timestamp` and `duration` have slightly similar semantics,
+and they define properties under the `composite` context.
+
+- `timestamp`: The start timestamp of the first span.
+- `duration`: The sum of durations of all spans.
+- `composite`
+    - `count`: The number of compressed spans this composite span represents.
+      The minimum count is 2 as a composite span represents at least two spans.
+    - `end`: The end timestamp of the last compressed span.
+      The net duration of all compressed spans is equal to the composite spans' `duration`.
+      The gross duration (including "whitespace" between the spans) is equal to `compressed.end - timestamp`.
```

Comment:
I wonder if it's worth going with the same model we have in breakdown metrics

- `sum.us` (a new property in `composite` object) to hold the sum of durations of all compressed spans
- `duration` holding gross duration (i.e., <last compressed span's end timestamp> - <first compressed span's start timestamp>)
- `end` being unnecessary anymore (effectively being replace by `sum.us`)

The pros of the new model are:
- `sum.us` and `count` is what we already have in breakdown metrics so it's more consistent
- Compressed-spans unaware UI will display a composite span using its gross duration which is closer to what we want although that might be debatable :)

WDYT?

Reply:
I'm not sure if we'd gain a lot there. If an old UI visualizes a composite span it will always display one thing wrong - either the net duration or the total duration. I'm not convinced that showing the gross duration is better than showing the net duration.

Reply:
> I'm not sure if we'd gain a lot there. If an old UI visualizes a composite span it will always display one thing wrong - either the net duration or the total duration. I'm not convinced that showing the gross duration is better than showing the net duration.

Yes, I agree the advantage to consider is the first one -  the consistency with breakdown metrics.

Reply:
I don't have a strong opinion on that. But I don't think we need to conform to metric naming conventions on things that are not metrics.
(With metrics, I'm specifically referring to the periodically flushed metrics that are aggregated in-memory in a metrics registry.)